mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,317 @@
+//! This module offers a transport layer for managing request-response style communication
+//! with Ethereum nodes using websockets in a wait and lock-free manner. In comparison to
+//! HTTP transport, this approach proves to be much quicker (low-latency) and consumes less
+//! bandwidth. This efficiency is achieved by avoiding the handling of TCP
+//! handshakes (connection reusability) for each request.
+
+use super::handle_gui_auth_payload;
+use super::http_transport::de_rpc_response;
+use crate::eth::web3_transport::Web3SendOut;
+use crate::eth::{EthCoin, RpcTransportEventHandlerShared};
+use crate::{MmCoin, RpcTransportEventHandler};
+use common::executor::{AbortSettings, SpawnAbortable, Timer};
+use common::expirable_map::ExpirableMap;
+use common::log;
+use futures::channel::mpsc::{UnboundedReceiver, UnboundedSender};
+use futures::channel::oneshot;
+use futures::lock::Mutex as AsyncMutex;
+use futures_ticker::Ticker;
+use futures_util::{FutureExt, SinkExt, StreamExt};
+use instant::Duration;
+use jsonrpc_core::Call;
+use mm2_net::transport::GuiAuthValidationGenerator;
+use std::collections::HashMap;
+use std::sync::{atomic::{AtomicUsize, Ordering},
+                Arc};
+use web3::error::{Error, TransportError};
+use web3::helpers::to_string;
+use web3::{helpers::build_request, RequestId, Transport};
+
+const REQUEST_TIMEOUT_AS_SEC: u64 = 10;
+
+#[derive(Clone, Debug)]
+pub(crate) struct WebsocketTransportNode {
+    pub(crate) uri: http::Uri,
+    pub(crate) gui_auth: bool,
+}
+
+#[derive(Clone, Debug)]
+pub struct WebsocketTransport {
+    request_id: Arc<AtomicUsize>,
+    node: WebsocketTransportNode,
+    event_handlers: Vec<RpcTransportEventHandlerShared>,
+    pub(crate) gui_auth_validation_generator: Option<GuiAuthValidationGenerator>,
+    responses: Arc<SafeMapPtr>,
+    controller_channel: Arc<ControllerChannel>,
+    connection_guard: Arc<AsyncMutex<()>>,
+}
+
+#[derive(Debug)]
+struct ControllerChannel {
+    tx: Arc<AsyncMutex<UnboundedSender<ControllerMessage>>>,
+    rx: Arc<AsyncMutex<UnboundedReceiver<ControllerMessage>>>,
+}
+
+enum ControllerMessage {
+    Request(WsRequest),
+    Close,
+}
+
+#[derive(Debug)]
+struct WsRequest {
+    serialized_request: String,
+    request_id: RequestId,
+    response_notifier: oneshot::Sender<()>,
+}
+
+/// A wrapper type for raw pointers used as a mutable Send & Sync HashMap reference.
+///
+/// Safety notes:
+///
+/// The implemented algorithm for socket request-response is already thread-safe,
+/// so we don't care about race conditions.
+///
+/// As for deallocations, see the `Drop` implementation below.
+#[derive(Debug)]
+struct SafeMapPtr(*mut HashMap<RequestId, Vec<u8>>);
+
+impl Drop for SafeMapPtr {
+    fn drop(&mut self) {
+        // Let the compiler do the job.
+        let _ = unsafe { Box::from_raw(self.0) };
+    }
+}
+
+unsafe impl Send for SafeMapPtr {}
+unsafe impl Sync for SafeMapPtr {}
+
```

Comment:
I believe we should consider using `parking_lot::Mutex` instead using `raw ptr` of writing/using unsafe code blocks, especially if the performance difference between the two isn't significant. Personally, I find unsafe code unpleasant :), so using a safer alternative like `parking_lot::Mutex` might preferable

Reply:
Almost the whole point of this algorithm is being lock and wait free.

https://github.com/KomodoPlatform/komodo-defi-framework/blob/febad662aef10c956824e31f30a84f0d30ec3db5/mm2src/coins/eth/web3_transport/websocket_transport.rs#L1-L2

https://github.com/KomodoPlatform/komodo-defi-framework/blob/febad662aef10c956824e31f30a84f0d30ec3db5/mm2src/coins/eth/web3_transport/websocket_transport.rs#L67-L83

Reply:
okay, can you add to the doc that the module uses unsafe code?